src/utils/cudnn.cpp

```diff
@@ -528,6 +528,332 @@ cudnnTensorDescriptor_t& entrywise_layer_tensor_manager::get_error_signals(int p
   return desc;
 }
 
+////////////////////////////////////////////////////////////
+// cuDNN algorithm selection
+////////////////////////////////////////////////////////////
+
+namespace {
+
+// Non-deterministic algorithms.
+std::vector<cudnnConvolutionFwdAlgo_t> nondet_fwd_algos = {};
+std::vector<cudnnConvolutionBwdDataAlgo_t> nondet_bwd_data_algos = {
+  CUDNN_CONVOLUTION_BWD_DATA_ALGO_0
+};
+std::vector<cudnnConvolutionBwdFilterAlgo_t> nondet_bwd_filter_algos = {
+  CUDNN_CONVOLUTION_BWD_FILTER_ALGO_0,
+  CUDNN_CONVOLUTION_BWD_FILTER_ALGO_3
+};
+
+template <typename AlgoType, typename PerfType>
+AlgoType find_best_heuristic_algorithm(
+  const std::vector<PerfType>& perf_results,
+  const std::vector<AlgoType>& deterministic_algos,
+  bool deterministic,
+  size_t max_ws_size) {
+  std::vector<AlgoType> algos;
+  for (const auto& p : perf_results) {
+    if (p.status != CUDNN_STATUS_SUCCESS) {
+      continue;
+    }
+    if (deterministic &&
+        std::find(deterministic_algos.begin(), deterministic_algos.end(),
+                  p.algo) != deterministic_algos.end()) {
+      continue;
+    }
+    if (p.memory > max_ws_size) {
+      continue;
+    }
+    algos.push_back(p.algo);
+  }
+  if (algos.empty()) {
+    LBANN_ERROR("No valid convolution algorithms.");
+  }
+  return algos[0];
+}
+
+template <typename AlgoType, typename PerfType>
+AlgoType find_best_algorithm(
+  const std::vector<PerfType>& perf_results,
+  const std::vector<AlgoType>& deterministic_algos,
+  bool deterministic,
+  size_t max_ws_size) {
+  std::unordered_map<AlgoType, float> time_map;
+  for (const auto& p : perf_results) {
+    if (p.status != CUDNN_STATUS_SUCCESS) {
+      continue;
```

Comment:
If an algorithm fails some runs and passes in others, then it's runtime will appear smaller than expected. Perhaps failure should add a runtime of infinity to the runtime.